aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,389 @@
+    /// 
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn generate(alg: KemAlgorithm) -> Result<Self, Unspecified> {
+        match alg {
+            KemAlgorithm::KYBER512_R3 => unsafe {
+                let kyber_key = kem_key_generate(alg.nid())?;
+                let mut priv_key_bytes = vec![0u8; alg.get_privkey_len()];
+                if 1 != EVP_PKEY_get_raw_private_key(*kyber_key, priv_key_bytes.as_mut_ptr(), &mut alg.get_privkey_len()) {
+                    return Err(Unspecified);
+                }
+                Ok(KemPrivateKey {
+                    algorithm: alg,
+                    context: LcPtr::from(kyber_key),
+                    priv_key: priv_key_bytes
+                })
+            },
+        }
+    }
+
+    /// Return the algorithm associated of the given KEM private key.
+    pub fn algorithm(&self) -> &KemAlgorithm {
+        &self.algorithm
+    }
+
+    /// Computes the KEM public key from the KEM private key
+    /// 
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    /// 
+    pub fn compute_public_key(&self) -> Result<KemPublicKey, Unspecified> {
+        let mut pubkey_bytes = vec![0u8; self.algorithm.get_pubkey_len()];
+        let mut pubkey_len;
+        let pubkey_ctx_copy;
+        match self.algorithm {
+            KemAlgorithm::KYBER512_R3 => {
+                pubkey_len = KYBER512_PUBLICKEYBYTES;
+            }
+        }
+        unsafe {
+            if 1 != EVP_PKEY_get_raw_public_key(*self.context, pubkey_bytes.as_mut_ptr(), &mut pubkey_len) {
+                return Err(Unspecified);
+            }
+            pubkey_ctx_copy = EVP_PKEY_kem_new_raw_public_key(self.algorithm.nid(), pubkey_bytes.as_mut_ptr(), pubkey_len);
+            if pubkey_ctx_copy.is_null() {
+                return Err(Unspecified);
+            }
+        }
+        Ok(KemPublicKey{
+            algorithm: self.algorithm.clone(),
+            context: LcPtr::new(pubkey_ctx_copy)?,
+            pub_key: pubkey_bytes
+        })
+    }
+
+    /// Performs the decapsulate operation using the current KEM private key on the given ciphertext.
+    /// 
+    /// `ciphertext` is the ciphertext generated by the encapsulate operation using the KEM public key
+    /// associated with the current KEM private key.
+    /// 
+    /// After the decapsulation is finished, `decapsulate` calls `kdf` with the raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    /// 
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    /// 
+    pub fn decapsulate<F, R>(&self, ciphertext: &mut [u8], kdf: F) -> Result<R, Unspecified>
+    where
+        F: FnOnce(&[u8]) -> Result<R, Unspecified> {
+        unsafe {
+            let ctx = DetachableLcPtr::new(EVP_PKEY_CTX_new(*self.context, null_mut()))?;
```

Comment:
I think all the places using `DetachableLcPtr` could just be `LcPtr`.

Reply:
Ack